[TOPI] [Relay] Sparse Conv2d Implementation for 3x3 kernels

Tantalus13A98B5F opened the pull request:
This PR currently comes with,

- NHWC & NCHW computes & schedules implemented for x86, by AutoTVM templates;
- Relay integration, reusing the infra contributed by existing 1x1 sparse conv2d kernels;
- A new relay pass to convert sparse models from models with freezed parameters;
- Tests not included for now; need discussion on coverage design;

Sample usages [here](https://github.com/Tantalus13A98B5F/convbench/blob/main/resnet50_conv3x3/tvm/integration.ipynb). Tested on Intel Xeon 6248 using ResNet18, single-threaded, we have the following results,

| Sparsity | Op Shape (NCHW) |                 |                 |               | ResNet18 Infer |
|:--------:|:---------------:|:---------------:|:---------------:|:-------------:|:--------------:|
|          |  10, 64, 56, 56 | 10, 128, 28, 28 | 10, 256, 14, 14 | 10, 512, 7, 7 |                |
|   Dense  | 0.0113          | 0.0107          | 0.0107          | 0.0127        | 0.2144         |
|    90%   | 0.0032          | 0.0031          | 0.0029          | 0.0032        | 0.0962         |
|    80%   | 0.0048          | 0.0049          | 0.0046          | 0.0053        | 0.1199         |
|    70%   | 0.0066          | 0.0065          | 0.0061          | 0.0065        | 0.1427         |
|    60%   | 0.0082          | 0.0076          | 0.0075          | 0.0078        | 0.1634         |
|    50%   | 0.0097          | 0.0085          | 0.0086          | 0.0092        | 0.1847         |
|    40%   | 0.0109          | 0.0097          | 0.0098          | 0.0107        | 0.1966         |
|    30%   | 0.0124          | 0.0110          | 0.0111          | 0.0122        | 0.2163         |
|    20%   | 0.0136          | 0.0123          | 0.0124          | 0.0138        | 0.2315         |
|    10%   | 0.0150          | 0.0137          | 0.0139          | 0.0154        | 0.2519         |

To conclude, acceleration is possible at the sparsity of 40%, compared with NCHWc convolutions.

@jcf94

junrushao1994 commented:
CC @yzh119 @MasterJH5574 @tkonolige @vinx13 if you guys are interested

XHPlus commented:
We also trained the relevant BSR models and achieved satisfactory accuracy. After completing the tuning, we will update the detailed accuracy and speed results. About the unit test coverage, we are looking forward to some suggestions for refinement. Thanks for the reviewers' efforts!

jcf94 commented:
Thanks! @Tantalus13A98B5F @XHPlus 
Also cc @yuchaoli

jcf94 commented on src/relay/transforms/convert_sparse_conv2d.cc:
```diff
@@ -155,6 +316,18 @@ Pass Conv2dToSparse(const Array<ObjectRef>& weight_name, const Array<Array<PrimE
 
 TVM_REGISTER_GLOBAL("relay._transform.Conv2dToSparse").set_body_typed(Conv2dToSparse);
 
+
+Pass Conv2dToSparse2(const String& layout, int kernel_size, int blockH, int blockW, double sparse_thresh) {
+  runtime::TypedPackedFunc<Function(Function, IRModule, PassContext)> pass_func =
+      [=](Function f, IRModule m, PassContext pc) {
+        auto f0 = Downcast<Function>(Conv2dToSparse2(f, layout, kernel_size, blockH, blockW, sparse_thresh));
+        return f0;
+      };
+  return CreateFunctionPass(pass_func, 5, "Conv2dToSparse2", {"DeadCodeElimination"});
+}
+
+TVM_REGISTER_GLOBAL("relay._transform.Conv2dToSparse2").set_body_typed(Conv2dToSparse2);
```
What's the different between the new added pass? Can this be merged to the original one?

Tantalus13A98B5F replied:
Sorry for the confusion. 

The original pass aims to transform models whose params are not binded, so that they still have names. Params are transformed to sparse formats prior to running the pass, and then pass is executed to replace convolutions with sparse counterparts, with the knowledge of names and shapes of sparse data.

The new pass aims to transform models with params binded. We found it useful when working with onnx, where specifying `freeze_params=True` is often desired. Data are inlined as `ConstantNode`s and have no names, thus param transformation and op replacement are both conducted in the new pass.

jcf94 commented on include/tvm/relay/attrs/nn.h:
```diff
@@ -1047,12 +1047,15 @@ struct SparseTransposeAttrs : public tvm::AttrsNode<SparseTransposeAttrs> {
 /*! \brief Attributes for sparse_dense operator */
 struct SparseConv2DAttrs : public tvm::AttrsNode<SparseConv2DAttrs> {
   std::string layout;
+  int kernel_size;
```
Suggest to follow the definition of `Conv2DAttrs` to use `Array<IndexExpr> kernel_size` here.

jcf94 commented on python/tvm/relay/analysis/sparse_conv2d.py:
```diff
@@ -108,19 +112,28 @@ def process_params(expr, params, block_size, sparsity_threshold, layout):
                 )
             else:
                 sparse_weight_data = sparse_weight.data
+        elif weight_kernel[0] == kernel_size == 3 and layout == "NHWC":
```
Only support NHWC here?

Tantalus13A98B5F replied:
Sorry to miss NCHW support, will be added in a later commit. In our current testing pipeline we don't use this to convert our models, though.

jcf94 replied:
Oh, that's fine, maybe add a TODO here or throw a warning on NCHW in case others have any confusion.

XHPlus commented:
![resnet18-tvm-record-core-i7](https://user-images.githubusercontent.com/13585124/128190694-4931014d-5150-45dd-a7b4-4efb6e50fe5a.png)

Hi everyone，we update the results with speed and accuracy for ResNet. (see the above figure)

We wonder whether there are some suggestions If we want to upload the relevant ONNX model and tuning log. Thanks!

Tantalus13A98B5F commented:
Update: addressed the issues above; added new tests for sparse conv3x3.

However the `BUILD:QEMU` target failed. Having no idea about that.

@jcf94 Any suggestions?

XHPlus commented:
Hi, the CI semes to fail due to the network problem. Do you have any suggestions to bypass it? Pushing a fake commit?

mengbingrock commented:
Hi @Tantalus13A98B5F 
Your work is very helpful!
I'm evaluating the performance of sparse kernels. I noticed that PR currently only supports padding = 1 case if I'm right. I've been struggling to add padding = 0 support.
I've modified the conv defined in `spconv2d_3x3_nhwc(cfg, data, wdat, wind, wptr, layout="NHWC")` and I think it could produce the correct result, or at least its dimension is correct. But it reports dimension error else where.
I will appreciate it if any guidance could be provided.

Thank you

Tantalus13A98B5F commented:
Hi @mengbingrock 
Not sure how long have you been going, but `bsr_conv2d.convert2` is merely for graph & param transformation, and it assumes well-padded shapes (0 for 1x1 and 1 for 3x3). Maybe you will want to tweak the implementation of this.

`spconv2d_3x3_nhwc` is the target-specialized implementation for the operator. In TVM, for every operator, we will want to [register it with Relay](https://tvm.apache.org/docs/dev/how_to/relay_add_op.html) so that we can use the operator in a graph. For the case of sparse operators, we would also like to provide transformations like `bsr_conv2d.convert2` to generate sparse graphs out of dense graphs. If we want our implementation to be visible for users, we have to fix all the code along the way.

For early-stage testing purposes, it's not necessary to do all the stuff at once. You can directly pass ndarrays to a scheduled template; if you have finished the Relay integration you can then manually compose a sparse graph for testing; once you've done the former steps you can consider tweaking the convert function.
